pkg/controllers/management/cluster/clustercontroller.go

```diff
@@ -0,0 +1,121 @@
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+type controller struct {
+	clusterClient v3.ClusterInterface
+	clusterLister v3.ClusterLister
+	nodeLister    v3.NodeLister
+}
+
+func Register(management *config.ManagementContext) {
+	c := controller{
+		clusterClient: management.Management.Clusters(""),
+		clusterLister: management.Management.Clusters("").Controller().Lister(),
+		nodeLister:    management.Management.Nodes("").Controller().Lister(),
+	}
+
+	c.clusterClient.AddHandler("clusterCreateUpdate", c.clusterSync)
+}
+
+func (c *controller) clusterSync(key string, obj *v3.Cluster) error {
+	var updated bool
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return nil
+	}
+
+	cluster, err := c.clusterClient.Get(key, metav1.GetOptions{})
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	}
+
+	if k8serrors.IsNotFound(err) {
+		return nil
+	}
+
+	if cluster.Status.Driver == v3.ClusterDriverImported {
+		return nil
+	}
+	capabilities := v3.Capabilities{}
+	capabilities.NodePortRange = "30000-32767"
+
+	if cluster.Spec.GoogleKubernetesEngineConfig != nil {
+		capabilities.L4LoadBalancer = c.L4Capability(true, "GCLB", []string{"TCP", "UDP"}, true)
+		if *cluster.Spec.GoogleKubernetesEngineConfig.EnableHTTPLoadBalancing {
+			capabilities.IngressController = c.IngressCapability(true, "GCLB", true)
+		}
+		updated = true
+	}
+	if cluster.Spec.AmazonElasticContainerServiceConfig != nil {
+		capabilities.L4LoadBalancer = c.L4Capability(true, "ELB", []string{"TCP"}, true)
+		updated = true
+	}
+	if cluster.Spec.AzureKubernetesServiceConfig != nil {
+		capabilities.L4LoadBalancer = c.L4Capability(true, "Azure L4 LB", []string{"TCP", "UDP"}, true)
+		updated = true
+		// on AKS portal you can enable Azure HTTP Application routing but Rancher doesn't have that option yet
+	}
+	if cluster.Spec.RancherKubernetesEngineConfig != nil {
+		switch cluster.Spec.RancherKubernetesEngineConfig.CloudProvider.Name {
+		case aws.AWSCloudProviderName:
+			capabilities.L4LoadBalancer = c.L4Capability(true, "ELB", []string{"TCP"}, true)
+		case azure.AzureCloudProviderName:
+			capabilities.L4LoadBalancer = c.L4Capability(true, "Azure L4 LB", []string{"TCP", "UDP"}, true)
+		}
+		// only if not custom, non custom clusters have nodepools set
+		nodes, err := c.nodeLister.List(cluster.Name, labels.Everything())
+		if err != nil {
+			return err
+		}
+
+		if len(nodes) > 0 {
+			if nodes[0].Spec.NodePoolName != "" {
+				capabilities.NodePoolScalingSupported = true
+			}
+		}
+
+		capabilities.IngressController = c.IngressCapability(true, "Nginx", false)
+		if cluster.Spec.RancherKubernetesEngineConfig.Services.KubeAPI.ExtraArgs["service-node-port-range"] != "" {
+			capabilities.NodePortRange = cluster.Spec.RancherKubernetesEngineConfig.Services.KubeAPI.ExtraArgs["service-node-port-range"]
+		}
+		updated = true
+	}
+
+	if updated {
```

Comment:
we should determine the update purely by comparing capabilities field of existing cluster with newly generated capabilities